# Tasks 



In [2]:
# Imports
import pandas as pd
import numpy as np
import sklearn as skl

### Task 1: Source the Data Set


In [3]:
# import the Iris data set from the sklearn.datasets module.
from sklearn.datasets import load_iris

# load the iris dataset
iris = load_iris()


#### 'Explain, in your own words, what the `load_iris()` function returns.'

The Iris data set is a data set which collects a number of samples from different species of irises, and is often used in statistics and machine learning, as mentioned [**here**](https://archive.ics.uci.edu/dataset/53/iris) by a paper published by **UC Irvine**.

Running the `load_iris()` function returns a Bunch object, which contains a data array, and all of the necessary keys to use the data set efficiently. This function will now allow me to explore this dataset, which you can see me do below in **Task 2**. 



### Task 2: Explore the Data Structure

#### 'Print and explain the shape of the data set, the first and last 5 rows of the data, the feature names, and the target classes.'

In [4]:
# print the data shape
print(iris.data.shape)

(150, 4)


The output shows us `(150, 4)`, which is a NumPy array of the data set. 150, being the 50 instances of each Iris class, mentioned previously in the UCI paper, whereas 4 is the number of characteristics measured from each sample. Let's explore further into the shape of the data set now using NumPy.

In [5]:
# convert data into a NumPy array
iris_data = np.array(iris.data)

# print the first and last 5 rows
print("The first 5 rows:\n", iris_data[:5])

print("The last 5 rows:\n", iris_data[-5:]) # negative indexing more easily returns the final 5 rows

The first 5 rows:
 [[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]
The last 5 rows:
 [[6.7 3.  5.2 2.3]
 [6.3 2.5 5.  1.9]
 [6.5 3.  5.2 2. ]
 [6.2 3.4 5.4 2.3]
 [5.9 3.  5.1 1.8]]


Now that we have the NumPy array, let's use Pandas to make it more readable!

In [8]:
# show target names
iris_data['target_names']


IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

In [8]:


# create dataframe with feature names
df = pd.DataFrame(iris.data, columns=iris.feature_names)

# add labels to dataframe
df['target'] = iris.target



Feature names: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
Keys: dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


### Task 3: Summarize the Data

For each feature in the dataset, calculate and display:

- mean
- minimum
- maximum
- standard deviation
- median


### Task 4: Visualize Features

Plot histograms for each feature using `matplotlib`.

Add appropriate titles and axis labels.

### Task 5: Investigate Relationships

Choose any two features from the data set and create a scatter plot of them.

Color-code the three different classes of the scatter plot points.

### Task 6: Analyze Relationship

Use `numpy.polyfit` to add a regression line to the scatter plot from Task 5.

### Task 7: Analyze Class Distributions

Create box-plots of the petal lengths for each of the three classes.

### Task 8: Compute Correlations

Calculate the correlation coefficients between the features.

Display the results as a heatmap using `matplotlib`.

### Task 9: Fit a Simple Linear Regression

For your two features in Task 5, calculate the coefficient of determination $R^2$.

Re-create the plot from Task 6 and annotate it with the $R^2$ value.

### Task 10: Too Many Features

Use `seaborn` to create a `pairplot` of the data set.

Explain, in your own words, what the `pairplot` depicts.